R/relativeOverfitting.R

```diff
@@ -19,37 +23,64 @@
 }
 
 #' @export
-#' @rdname estimateRelativeOverfitting
-estimateRelativeOverfitting.ResampleDesc = function(rdesc, measures, task, learner) {
+estimateRelativeOverfitting.ResampleDesc = function(predish, measures, task, learner, ...) {
+  assertClass(predish, classes = "ResampleDesc")
+  assertClass(learner, classes = "Learner")
   measures = checkMeasures(measures, task)
 
-  rdesc$predict = "both"
-  r = resample(learner, task, rdesc, measures, show.info = FALSE)
+  predish$predict = "both"
+  r = resample(learner, task, predish, measures, show.info = FALSE)
+
+  estimateRelativeOverfitting(r$pred, measures, task)
+}
+
+#' @export
+estimateRelativeOverfitting.ResamplePrediction = function(predish, measures, task, ...) {
+  assertClass(predish, classes = "ResamplePrediction")
+  measures = checkMeasures(measures, task)
   mids = vcapply(measures, function(m) m$id)
 
-  iterations = nrow(r$measures.test)
-  do.call(rbind, lapply(1:iterations, function(i) {
-    perf.test = r$measures.test[i,mids,drop = FALSE]
-    perf.train = r$measures.train[i,mids,drop = FALSE]
-
-    data = r$pred$data[r$pred$data$iter == i & r$pred$data$set == "test",]
-    nrows = nrow(data)
-    pred.permuted = r$pred
-    pred.permuted$data = data.frame(truth = rep(data$truth, each = nrows),
-      response = rep(data$response, times = nrows))
-    perf.permuted = performance(pred.permuted, measures = measures, task = task)
-
-    df = (perf.test - perf.train) / (perf.permuted - perf.train)
-    names(df) = stri_paste("relative.overfit", mids, sep = ".")
-    cbind(data.frame(iter = i), df)
+  iterations = unique(predish$data$iter)
+  rbindlist(lapply(iterations, function(i) {
+    data = predish$data[predish$data$iter == i & predish$data$set == "test",]
+    pred.test = makePrediction(task$task.desc, row.names(data), data$id, data$truth, predish$predict.type, predish$predict.threshold, data$response, predish$time[i])
+
+    data = predish$data[predish$data$iter != i & predish$data$set == "test",]
+    pred.train = makePrediction(task$task.desc, row.names(data), data$id, data$truth, predish$predict.type, predish$predict.threshold, data$response, predish$time[i])
+
+    estimateRelativeOverfitting(pred.test, measures, task, pred.train = pred.train, iter = i)
   }))
 }
+
+#' @export
+estimateRelativeOverfitting.Prediction = function(predish, measures, task, learner, pred.train, iter = 1) {
+  assertClass(predish, classes = "Prediction")
+  assertClass(pred.train, classes = "Prediction")
+  measures = checkMeasures(measures, task)
+  mids = vcapply(measures, function(m) m$id)
+
+  perf.test = performance(predish, measures = measures)
+  perf.train = performance(pred.train, measures = measures)
+
+  nrows = nrow(predish$data) + nrow(pred.train$data)
+  pred.permuted = predish
+  pred.permuted$data = data.frame(truth = rep(c(predish$data$truth, pred.train$data$truth), each = nrows),
+    response = rep(c(predish$data$response, pred.train$data$response), times = nrows))
```

Comment:
(Without having a deeper look into the objects): I suspect `predish$data` and `pred.train$data` have the same number of rows -- why do you need to call `rep` once with `each` and once with `times`? Maybe a short comment would clarify this line.